File: Python/Object Oriented Design/JukeBox.py
# Design a musical jukebox using object oriented design


class CD(object):
    def __init__(self, cd_name=None, *songs):
        self.cd_name = cd_name
        self.songs = songs


class CDPlayer(object):
    def __init__(self):
        self.max_number_of_cds = 10
        self.cd_list = []

    def add_cd(self, cd_name, *songs):
        if len(self.cd_list) < 10:
            self.cd_list.append(CD(cd_name, songs))
        else:
            print "CD Player is full, try removing a cd first"

    def remove_cd(self, cd_name):
        for i in xrange(len(self.cd_list)):
            try:
                cd_obj = self.cd_list[i]
                if cd_name == cd_obj.cd_name:
                    del self.cd_list[i]
            except IndexError:
                pass

    def clear_cd_list(self):
        self.cd_list = []

    def get_cd_list(self):
        return self.cd_list


class User(CDPlayer):
    def __init__(self, name, user_id):
        super(User, self).__init__()
        self.name = name
        self.ID = user_id
        self.users = [self.name + self.ID]

    def get_users(self):
        return self.users

    def add_user(self, name, user_id):
        self.users.append(name + user_id)

    def delete_user(self, name, user_id):
        self.users.remove(name + user_id)


class JukeBox(User):
    def __init__(self, name, user_id):
        super(JukeBox, self).__init__(name, user_id)
        self.default_user = User.get_users(self)[0]
        self.save_state = {}

    def change_user(self, name, user_id):
        u = User.get_users(self)
        for i in xrange(len(u)):
            if name + user_id == u[i]:
                self.default_user = u[i]
        CDPlayer.clear_cd_list(self)

    def save_user_music_list(self):
        self.save_state[self.default_user] = CDPlayer.get_cd_list(self)

    def display_user_information(self):
        values = self.save_state[self.default_user]
        print "user: " + self.default_user
        for i in values:
            print "Album:", i.cd_name, ", Song List:", \
                str(i.songs).replace("(", "").replace(")", "").replace("[", "").replace("]", "").rstrip(",")


    # Other things I might add
    # def -> the ability to change cds next and prev

def main():
    jb = JukeBox("John", "123")
    jb.add_cd("Kenny Bloom", ["burn burn burn", "train", "grey hat"])
    jb.add_cd("Betsy Land", ["I love candy", "objects are great", "please inherit me"])
    jb.save_user_music_list()
    jb.display_user_information()

    jb.add_user("Doe", "115")
    jb.change_user("Doe", "115")
    jb.add_cd("Benny Boo", ["falling asleep", "tip toe", "walking to town"])
    jb.add_cd("Mikey Kite", ["ding dong", "flying through walls", "faking it"])
    jb.save_user_music_list()
    jb.display_user_information()
    print


if __name__ in "__main__":
    main()

File: Java/Hired Programming Practice/src/ObjectOrientedDesign/JukeBox.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ObjectOrientedDesign;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 *
 * @author marcsantiago
 */
// Design a musical jukebox using object oriented design


class CD{
    private final String cdName;
    private final String [] songs;
    private final String album;
    public CD(String cdName, String [] songs){
        this.cdName = cdName;
        this.songs = songs;
        this.album = cdName + " " + Arrays.toString(songs);
    }
    
    public String getAlbum(){
        return album;
    }
    
    public String getCDName(){
        return cdName;
    }
}

class CDPlayer{
    private final int maxCDS = 10;
    private final ArrayList<CD> cdList = new ArrayList<>();
    
    public void addCD(String cdName, String [] songs){
        if(cdList.size() < 10){
            cdList.add(new CD(cdName, songs));
        }
        else{
            System.out.println("CD player is full try removing a cd first");
        }
    }
    
    public void removeCD(String cdName){
        for (int i = 0; i < cdList.size(); i++) {
            if(cdList.get(i).getCDName().equals(cdName)){
                cdList.remove(i);
            }
        }
    }
    
    public void clearCDList(){
        cdList.clear();
    }
    
    public ArrayList<CD> getCDList(){
        return cdList;
    }
}

class User extends CDPlayer{
    private String name = "";
    private String id = "";
    private final ArrayList<User> users = new ArrayList<>();
    
    public User(){

    }
 
    public User(String name, String id){
        this.name = name;
        this.id = id;
    }

    
    public void addUser(String name, String id){        
        User newPerson = new User(name, id);
        if(!users.contains(newPerson)){
            users.add(newPerson);
        }
    }
    
    public void deleteUser(String name, String id){
        try{
            users.remove(new User(name, id));
        }
        catch(Exception e){
            System.out.println(e);
        }
        
    }
    
    public ArrayList<User> getUsers(){
        return users;
    }
    
    public String getName(){
        return name;
    }
    
    public String getID(){
        return id;
    }
    
}

public class JukeBox extends User{
    private User defaultUser;
    private Map<User, ArrayList<CD> > saveState = new HashMap<>();
    public JukeBox(String name, String id) {
        addUser(name, id);
        defaultUser = new User(name, id);
    }   
    
    public void changeDefaultUser(String name, String id){  
        User newPerson = new User(name, id);
        for(User u : getUsers()){
            if(newPerson.getName().equals(u.getName())){
                this.defaultUser = newPerson;
            }
        }
        clearCDList();
    }
    
    public void saveUserMusic(){
        saveState.put(defaultUser, getCDList());
    }
    
    public void displayCurrentUserInformation(){
        if(!saveState.isEmpty()){
            System.out.println("User: " + defaultUser.getName() + " ID: " + defaultUser.getID());
            ArrayList<CD> v  = saveState.get(defaultUser);
            for(CD c : v){
                System.out.println(c.getCDName() + " "  + c.getAlbum());
            }     
        }
        else{
            System.out.println("No Data has been saved to display");
            System.out.println("Remeber to save user data");
        }
        
    }
}